In [1]:
import numpy as np

In [2]:
a = [1,4,5.5,"string", True]
a

[1, 4, 5.5, 'string', True]

In [3]:
np.array(a)         # numpy accepts homogenous kinds of elements

array(['1', '4', '5.5', 'string', 'True'], dtype='<U32')

In [4]:
b = [1,2,3,4,5,6]

np.array(b)

array([1, 2, 3, 4, 5, 6])

# -> Numpy is of fixed size, it can grow but create a completely new array, whereas list can grow dynamically
# -> List - overhead to store all memory locations. Numpy arrays only stores similar type of data

#### Numpy is faster

In [5]:
size = 1000

x1 = range(size)
x2 = range(size)

In [6]:
%%timeit

[x1[i] * x2[i] for i in range(size)]

1.42 ms ± 250 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [7]:
y1 = np.arange(size)
y2 = np.arange(size)

In [8]:
%%timeit

y1 * y2

4.29 µs ± 464 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


#### you could see the difference , Numpy is a lot faster then list in terms of time
##### the list took 1420 µs
##### the numpy took 4.29 µs

#  Views vs Copy

In [10]:
a = np.array([1,2,3,4,5])

In [11]:
b = a

In [12]:
a , b

(array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5]))

In [13]:
a[0] = -5

In [14]:
a , b              # both a and b changed to -5

(array([-5,  2,  3,  4,  5]), array([-5,  2,  3,  4,  5]))

In [18]:
c = a.view()

In [19]:
a , b , c

(array([-5,  2,  3,  4,  5]),
 array([-5,  2,  3,  4,  5]),
 array([-5,  2,  3,  4,  5]))

In [20]:
x = a.copy()

In [21]:
x[1] = 20

In [22]:
a , x      # different values now

(array([-5,  2,  3,  4,  5]), array([-5, 20,  3,  4,  5]))